### PR TITLE
Implementation of recommendation 3.7 - usage of spaces for code separation (parts 3 and 4)

### LINKED ISSUE #151: Recommendation 3.7: Usage of whitespace for code separation
1. Any keywords (like 'if', 'when', 'for', e.t.c) should be separated with a single whitespace from the opening parenthesis.
     Only exceptions are: 'super' and 'constructor' keywords. They should not be separated from a parenthesis.

  2. Separate any keywords (such as `else` or `catch`, etc.) from the opening brace ('{') with a single whitespace.
  
  3. Use single whitespace before any opening brace (`{`).
     Only exception is passing a lambda as an argument inside parenthesis:
     ```kotlin
         private fun foo(a: (Int) -> Int, b: Int) {}
         foo({x: Int -> x}, 5) // no space before '{'
     ```

  4. Single whitespace should be placed on both sides of binary operators. This also applies to operator-like symbols, for example: 

     - Colon in generic structures with 'where' keyword： `where T : Type`
     - Arrow in lambdas： `(str: String) -> str.length()`

  **Exceptions：**

  - Two colons (`::`) are written without spaces: `Object::toString`
  - Dot separator (`.`) that stays on the same line with an object name: `object.toString()`

  5. Spaces should be used after ',',':',';' (except cases when those symbols are in the end of line). There should be no whitespaces in the end of line.

  6. There should be *only one space* between identifier and it's type： `list: List<String>`
  
  7. When using '[]' operator (get/set) there should be *no* spaces between identifier and '[': `someList[0]`
  
  8. There should be no space between a method name and a parenthesis: `foo() {}`

### PR DESCRIPTION
**Pull request by petertrr**
Until #166 is merged, this pull request is pointed in that branch.

## Which rule and warnings did you add?

This pull request *partially* closes #151 (parts 3 and 4)

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers

## TODO
* [ ] Safe access modifiers: `?.` and `!!`, that stay on the same line with an object name: `object?.toString()`

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt [hunk 1]
```diff
@@ -2,33 +2,60 @@ package org.cqfn.diktat.ruleset.rules
 
 import com.pinterest.ktlint.core.KtLint
 import com.pinterest.ktlint.core.Rule
+import com.pinterest.ktlint.core.ast.ElementType.ARROW
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt:5]: I am starting to think that wildcard are not so bad idea...

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 2]
```diff
@@ -1091,6 +1091,7 @@ fun baz() {
 
   - Two colons (`::`) are written without spaces: `Object::toString`
   - Dot separator (`.`) that stays on the same line with an object name `object.toString()`
+  - Operator `..` for creating ranges, e.g. `1..100`
```
Comment (akuleshov7) [inline info/diktat-kotlin-coding-style-guide-en.md:1094]: compilation does not prohibit anything else?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt [hunk 3]
```diff
@@ -61,14 +90,88 @@ class WhiteSpaceRule : Rule("horizontal-whitespace") {
...
-    private fun ASTNode.numWhiteSpaces() = if (elementType == WHITE_SPACE) text.count { it == ' ' } else 0
+    /**
+     * This method covers all other opening braces, not covered in [handleKeywordWithParOrBrace].
+     */
+    private fun handleLbrace(node: ASTNode) {
+        val whitespaceOrPrevNode = node.parent({ it.treePrev != null }, strict = false)!!.treePrev
+        val isFromLambdaAsArgument = node.parents().take(numParentsForLambda).toList().let {
+            it[0].elementType == FUNCTION_LITERAL &&
+                    it[1].elementType == LAMBDA_EXPRESSION &&
+                    it[2].elementType == VALUE_ARGUMENT
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt:114]: don't you think that you can get index out of bound in some cases here?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt (hunk 1 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt:5]: Maybe, we should support exceptions to several wildcards through configuration, like
```
   allowedWildcards="com.pinterest.ktlint.core.ast.*,..."
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt:114]: Right, I will add a check.

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 2 above)
Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:1094]: I have checked, things like `1 .. 100` are allowed, you can even insert newline after the `..`.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt (hunk 1 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhiteSpaceRule.kt:5]: Created #185

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 4]
```diff
@@ -1294,7 +1320,7 @@ And better use other names instead of these identifiers.
 ### <a name="r3.8"></a>Rule 3.8: Concatenation Strings is prohibited when string fits one line, use raw strings and string templates instead.
 Kotlin significantly enhanced work with Strings:
 [String templates](https://kotlinlang.org/docs/reference/basic-types.html#string-templates), [Raw strings](https://kotlinlang.org/docs/reference/basic-types.html#string-literals)
-That's why code looks much better when instead of using explicit concatenation to use proper Kotlin strings in case your line is not too long and you do not need to split it with newlines.
+That's why code looks much better when instead of using explicit concatenation to use proper Kotlin strings.
```
Comment (akuleshov7) [inline info/diktat-kotlin-coding-style-guide-en.md:1323]: fix back

**Pull request merged**